### PR TITLE
[TextPainter] Don't invalidate layout cache for paint only changes

### LINKED ISSUE #85108: Unhandled Exception: 'package:flutter/src/painting/text_painter.dart': Failed assertion: line 881 pos 12: '!_needsLayout': is not true.
labelText breaks TextField with the assertion !_needsLayout

##### Minimal example: 

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }
  OutlineInputBorder border = OutlineInputBorder(
    borderRadius: BorderRadius.circular(30),
  );
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: TextField(
          cursorHeight: 16,
          cursorColor: Colors.red,
          decoration: InputDecoration(
            contentPadding: const EdgeInsets.only(left: 24, top: 34, right: 24),
            fillColor: Colors.white,
            filled: true,
            labelText: "@something",
            floatingLabelBehavior: FloatingLabelBehavior.never,
            errorBorder: border,
            enabledBorder: border,
            focusedBorder: border,

          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
``` 

</details>

##### Steps to reproduce:
1. Create MacOS app
2. Focus on TextField
3. Wiggle mouse around where the label was, try go off window and off TextField for maximum effect 
4. Remove labelText and observe bug disappears

##### Flutter doctor

<details> 
<summary>flutter doctor -v</summary> 
	
```console
❯ flutter doctor -v
[✓] Flutter (Channel stable, 2.2.2, on macOS 11.4 20F71 darwin-x64, locale en-GB)
    • Flutter version 2.2.2 at /Users/mousybusiness/Flutter/flutter
    • Framework revision d79295af24 (12 days ago), 2021-06-11 08:56:01 -0700
    • Engine revision 91c9fc8fe0
    • Dart version 2.13.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/mousybusiness/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.4 20F71 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 91.0.4472.114

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
- Fixes https://github.com/flutter/flutter/issues/85108

Defers the rebuilding and the relayout of the ui.Paragraph object to
`paint` if there are no layout changes (e.g. only the font color
changed) and the width constraints are the same.

Additionally if the `TextPainter` is used to help compute the dry layout
of a render object, and thus never needs to paint, then color changes
alone no longer triggers relayout.

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -185,8 +185,12 @@ class TextPainter {
        _textWidthBasis = textWidthBasis,
        _textHeightBehavior = textHeightBehavior;
 
+  // _paragraph being null means the text needs layout because of style changes.
+  // Setting _paragraph to null invalidates all the layout cache.
   ui.Paragraph? _paragraph;
-  bool _needsLayout = true;
+  // Whether _paragraph contains outdated paint information and needs to be
+  // rebuilt.
+  bool _needsPaint = true;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:193]: It's a little confusing that the property is called needsPaint, but the comment talks about needing rebuild...

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 2]
```diff
@@ -561,38 +576,29 @@ class TextPainter {
...
-      _text!.build(builder, textScaleFactor: textScaleFactor, dimensions: _placeholderDimensions);
-      _inlinePlaceholderScales = builder.placeholderScales;
-      _paragraph = builder.build();
+  // Creates a ui.Paragraph using the current configurations in this class and
+  // assign it to _paragraph.
+  void _createParagraph() {
+    assert(_paragraph == null || _needsPaint);
+    final InlineSpan? text = this.text;
+    if (text == null) {
+      throw StateError('TextPainter.text must be set to a non-null value before using the TextPainter.');
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:592]: Why not an assert?

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 3]
```diff
@@ -630,15 +662,23 @@ class TextPainter {
...
+    final double? minWidth = _lastMinWidth;
+    final double? maxWidth = _lastMaxWidth;
+    if (_paragraph == null || minWidth == null || maxWidth == null) {
+      throw StateError(
+        'TextPainter.paint called when text geometry was not yet calculated.\n'
+        'Please call layout() before paint() to position the text before painting it.',
+      );
+    }
+
+    if (_needsPaint) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:674]: the name _needsPaint seems confusing to me. The TextPainter is drawing/painting the paragraph in line 682 even when this is false.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 4]
```diff
@@ -196,7 +200,7 @@ class TextPainter {
   /// in framework will automatically invoke this method.
   void markNeedsLayout() {
     _paragraph = null;
-    _needsLayout = true;
+    _needsPaint = true;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:203]: I believe, technically setting _paragraph to null here is probably enough?

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/painting/text_painter.dart:674]: _needsRebuildDuringPaint, maybe?

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 5]
```diff
@@ -496,23 +511,23 @@ class TextPainter {
   ///
   /// Valid only after [layout] has been called.
   double get minIntrinsicWidth {
-    assert(!_needsLayout);
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:499]: Maybe slightly more readable to define a getter `bool get _needsLayout => _paragraph != null` and use it here and elsewhere?

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 6]
```diff
@@ -630,15 +662,23 @@ class TextPainter {
...
+      );
+    }
+
+    if (_needsPaint) {
+      _createParagraph();
+      // Unfortunately we have to redo the layout using the same constraints,
+      // since we've created a new ui.Paragraph. But there's no extra work being
+      // done: if _needsPaint is true and _paragraph is not null, the previous
+      // `layout` call didn't invoke _layoutParagraph.
+      _layoutParagraph(minWidth, maxWidth);
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:680]: Can we assert that the size calculated by the new paragraph is the same as the size before?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:674]: `_rebuildParagraphForLayout` (a getter) and `_rebuildParagraphForPaint` ?

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/lib/src/painting/text_painter.dart:499]: `_paragraph` is declared nullable, so even with NNBD these could be null at this point and that would be a bug.

**Pull request merged**